text/0000-graceful-stack-upgrades.md

```diff
@@ -0,0 +1,118 @@
+# Meta
+
+[meta]: #meta
+
+- Name: Graceful stack upgrades
+- Start Date: 2022-07-12
+- Author(s): natalieparellano
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+
+[summary]: #summary
+
+This RFC proposes that the lifecycle skip layer re-use and clear its cache when it recognizes that the previous image
+was built with a different stack ID.
+
+# Definitions
+
+[definitions]: #definitions
+
+**lifecycle** - software that orchestrates a CNB build; it executes in a series of phases that each have a distinct
+responsibility.
+
+**layer re-use** - the lifecycle provides metadata about previously-built layers to buildpacks so that layers can be
+re-used if necessary - i.e., the lifecycle can avoid re-uploading unchanged layers to the registry and / or daemon. Only
+metadata about layers is copied from the previous image (not data).
+
+**cache** - layers that are not exported but may be needed again at build-time are saved in the cache.
+
+# Motivation
+
+[motivation]: #motivation
+
+Re-using layers that were built with a different underlying OS is unsafe. The layers may depend on OS-level dependencies
+that are not present in the new OS, causing errors that can only be seen at runtime. With Ubuntu 18.04 LTS (Bionic
+Beaver) [exiting the free initial five-year maintenance period](https://ubuntu.com/about/release-cycle) in April 2023,
+other Ubuntu versions - such as 20.04 LTS (Focal Fossa) and 22.04 LTS (Jammy Jellyfish) will become more prevalent in
+the CNB ecosystem. Accordingly, the risk is greater that some CNB-built images could break.
+
+Use case: as a platform operator, I want to be able to upgrade my stack to a newer Ubuntu release without worrying that
+my images will break.
+
+# How it Works
+
+[how-it-works]: #how-it-works
+
+The lifecycle when reading the previous image metadata will inspect the value of `io.buildpacks.stack.id`. If the value
+does not match the value of `CNB_STACK_ID` in the lifecycle's execution environment, the lifecycle will warn and skip
```

Comment:
When analyze/restore is running on the lifecycle image (as opposed to a trusted builder) do we expect that this env var is set? It looks like it may not be, [by pack at least](https://github.com/buildpacks/pack/blob/30de289455817bb90d760da0b900131443185a5d/internal/build/lifecycle_execution.go#L457), currently.

However, I don't know that we need it. We could instead use `io.buildpacks.stack.id` from the run image.

Reply:
Updated in 9590373